File: leetcode-daily/2025-08-aug/2025-08-10/2025-08-10-fun-with-linked-lists.c
// Title: Fun with linked lists
// Date: 2025-08-10
// Created: `Sun, 10 Aug 2025 20:34:24 +0530`
// Tags: [linked-lists, daily-practice]

#include <stdlib.h>

/// {{
/// Problem: 707. Design Linked List
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/design-linked-list/
/// Topics: `linked-list`, `design`
/// Timestamp: `Sun, 10 Aug 2025 20:34:24 +0530`

typedef int Val;

struct MyListNode {
  struct MyListNode *next;
  Val val;
};
typedef struct MyListNode *Node;

typedef struct {
  Node first;
  int size;
} MyLinkedList;

typedef MyLinkedList *List;

#define IS_LIST_EMPTY(list) (((List) list)->first == NULL)

const size_t LINKED_LIST_WIDTH = sizeof(MyLinkedList);
const size_t NODE_WIDTH = sizeof(struct MyListNode);

static Node getNodeAt(List list, int index);

List myLinkedListCreate();
void myLinkedListFree(List list);
 Val myLinkedListGet(List list, int index);
void myLinkedListAddAtHead(List list, Val val);
void myLinkedListAddAtTail(List list, Val val);
void myLinkedListAddAtIndex(List list, int index, Val val);
void myLinkedListDeleteAtIndex(List list, int index);

static Node getNodeAt(List list, int index) {
  if (!(0 <= index && index < list->size)) {
    return NULL;
  }

  if (index == 0) {
    return list->first;
  }

  Node node = list->first;

  for (int i = 1; i < (index + 1); i++) {
    if (node->next) {
      node = node->next;
    } else {
      return NULL;
    }
  }

  return node;
}

List myLinkedListCreate() {
  return calloc(1, LINKED_LIST_WIDTH);
}

void myLinkedListFree(List list) {
  while (!IS_LIST_EMPTY(list)) {
    myLinkedListDeleteAtIndex(list, 0);
  }
  free(list), (list = NULL);
}

int myLinkedListGet(List list, int index) {
  Node node = getNodeAt(list, index);
  return node ? node->val : -1;
}

inline void myLinkedListAddAtHead(List list, int val) {
  myLinkedListAddAtIndex(list, 0, val);
}

inline void myLinkedListAddAtTail(List list, int val) {
  myLinkedListAddAtIndex(list, list->size, val);
}

void myLinkedListAddAtIndex(List list, int index, int val) {
  if (!(0 <= index && index <= list->size)) {
    return;
  }

  Node cur = calloc(1, NODE_WIDTH);
  cur->val = val;

  Node prev = getNodeAt(list, index - 1);

  if (prev) {
    Node next = prev->next;
    prev->next = cur;
    cur->next = next;
    list->size++;
  } else {
    if (index == 0) {
      cur->next = list->first;
      list->first = cur;
      list->size++;
    }
  }
}

void myLinkedListDeleteAtIndex(List list, int index) {
  if (!(0 <= index && index < list->size)) {
    return;
  }
  Node prev = getNodeAt(list, index - 1);

  if (prev) {
    if (prev->next) {
      Node next = prev->next->next;
      free(prev->next), (prev->next = NULL);
      prev->next = next;
      list->size--;
    }
  } else {
    if (list->first) {
      Node next = list->first->next;
      free(list->first), (list->first = NULL);
      list->first = next;
      list->size--;
    }
  }
}

/**
 * Your MyLinkedList struct will be instantiated and called as such:
 * List list = myLinkedListCreate();
 * int param_1 = myLinkedListGet(obj, index);
 * myLinkedListAddAtHead(obj, val);
 * myLinkedListAddAtTail(obj, val);
 * myLinkedListAddAtIndex(obj, index, val);
 * myLinkedListDeleteAtIndex(obj, index);
 * myLinkedListFree(obj);
*/

/// }}

File: leetcode-daily/2025-08-aug/2025-08-10/2025-08-10-fun-with-linked-lists.js
// Title: Fun with linked lists
// Date: 2025-08-10
// Created: `Sun, 10 Aug 2025 22:12:24 +0530`
// Tags: [linked-lists, daily-practice]

/// {{
/// Problem: 707. Design Linked List
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/design-linked-list/
/// Topics: `linked-list`, `design`
/// Timestamp: `Sun, 10 Aug 2025 22:12:24 +0530`

/** @typedef {{next: ListNode | null; data: number}} ListNode */

function MyLinkedList() {
  /** @type {ListNode | null} */
  this.first = null;

  /** @type {number} */
  this.size = 0;
};

/**
 * @param {MyLinkedList} list 
 * @param {number} index
 * @returns {ListNode | null}
 */
function getNodeAt(list, index) {
  if (!list || !(0 <= index && index < list.size)) {
    return null;
  }

  if (index === 0) {
    return list.first;
  }

  /** @type {ListNode | null} */
  let node = list.first;

  for (let i = 1; i < (index + 1); i++) {
    if (node.next) {
      node = node.next;
    } else {
      return null;
    }
  }

  return node;
}

/** 
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
  const node = getNodeAt(this, index);
  return node ? node.data : -1;
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
  this.addAtIndex(0, val);
};

/** 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
  this.addAtIndex(this.size, val);
};

/** 
 * @param {number} index 
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {
  if (!(0 <= index && index <= this.size)) {
    return;
  }

  /** @type {ListNode} */
  const newNode = { next: null, data: val };
  const prev = getNodeAt(this, index - 1);

  if (prev) {
    const next = prev.next;
    prev.next = newNode;
    newNode.next = next;
    this.size++;
  } else {
    newNode.next = this.first;
    this.first = newNode;
    this.size++;
  }
};

/** 
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {
  if (!(0 <= index && index < this.size)) {
    return;
  }

  const prev = getNodeAt(this, index - 1);

  if (prev) {
    if (prev.next) {
      const next = prev.next.next;
      prev.next = next;
      this.size--;
    }
  } else {
    if (this.first) {
      this.first = this.first.next;
      this.size--;
    }
  }
};

/** 
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */

/// }}

